src/web/extractors.rs

```diff
@@ -42,44 +84,305 @@ impl FromRequest<ServerState> for MetaRequest {
+        } else {
+            // This ideally never happens, but just in case the location wasn't supplied
+            RequestErrorLocation::Unknown
+        };
+        RequestValidationError {
+            location,
+            name: name.to_string(),
+            description: format!("{}", error),
+        }
     }
 }
 
-#[derive(Deserialize)]
-pub struct CollectionParams {
-    pub uid: String,
-    pub collection: String,
+/// Validation Error Location in the request
+#[derive(Debug, Deserialize, Serialize, PartialEq)]
+#[serde(rename_all = "lowercase")]
+enum RequestErrorLocation {
+    Body,
+    QueryString,
+    Url,
+    Header,
+    Path,
+    Cookies,
+    Method,
+    Unknown,
 }
 
-#[derive(Deserialize)]
-pub struct BsoParams {
-    pub uid: String,
-    pub collection: String,
-    pub bso: String,
+/// Request Validation Errors
+///
+/// This is a wrapper on validator's ValidationErrors so that the error can
+/// be properly adapted to a ResponseError.
+#[derive(Debug)]
+pub struct RequestValidationErrors(ValidationErrors);
+
+impl error::Error for RequestValidationErrors {
+    fn description(&self) -> &str {
+        self.0.description()
+    }
+
+    fn cause(&self) -> Option<&error::Error> {
+        self.0.cause()
+    }
 }
 
-#[derive(Deserialize, Serialize)]
-pub struct BsoBody {
-    pub sortindex: Option<i32>,
-    pub payload: Option<String>,
-    pub ttl: Option<u32>,
+impl fmt::Display for RequestValidationErrors {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl ResponseError for RequestValidationErrors {
+    fn error_response(&self) -> HttpResponse {
+        let error: ClientRequestError = self.into();
+        HttpResponse::build(StatusCode::BAD_REQUEST).json(error)
+    }
+}
+
+impl<'a> From<&'a RequestValidationErrors> for ClientRequestError {
+    fn from(errors: &RequestValidationErrors) -> Self {
+        let mut sync_error = ClientRequestError::default();
+        sync_error.status = 400;
+        let mut client_errors = Vec::new();
+        let validation_errors = errors.0.clone().field_errors();
+        for (ref field_name, error_kind) in validation_errors.iter() {
+            for error in error_kind {
+                client_errors.push(RequestValidationError::with_name_and_error(
+                    field_name, error,
+                ))
+            }
+        }
+        if !client_errors.is_empty() {
+            sync_error.errors = Some(client_errors)
+        }
+        sync_error
+    }
+}
+
+/// Convenience function to create a `ValidationError` with additional context
+fn request_error(message: &'static str, location: RequestErrorLocation) -> ValidationError {
+    let mut err = ValidationError::new(message);
+    err.add_param("location".into(), &location);
+    err
+}
+
+/// Verifies that the supplied value is >= 0
+fn validate_qs_positive_value<T: PartialOrd + Zero>(ts: T) -> Result<(), ValidationError> {
```

Comment:
looks good though this module is already becoming pretty large (and with more to come). any pieces make sense to split off?

Reply:
Yup, definitely going to be splitting it up.